File: in_Java/0547 Friend Circles.java
class Solution {
    public int findCircleNum(int[][] M) {
        int num_nodes = M.length;
        int res = 0;
        boolean[] visited = new boolean[num_nodes];
        Stack<Integer> st = new Stack<>();
        int cur;
        for(int i = 0; i < num_nodes; i++){
            if(!visited[i]){
                res++;
                st.push(i);
                while(!st.isEmpty()){
                    cur = st.pop();
                    visited[cur] = true;
                    for(int j = 0; j < M[0].length; j++){
                        if(j != cur && M[cur][j] == 1){
                            if(!visited[j]) st.push(j);
                        }
                    }
                }
            }
        }
        return res;
    }
}

File: in_Python/0547 Friend Circles.py
class Solution(object):
    def findCircleNum(self, M):
        """
        :type M: List[List[int]]
        :rtype: int
        """
        visited = []
        res = 0
        for i in range(len(M)):
            if i in visited:
                continue
            q = []
            q.append(i)
            visited.append(i)
            while len(q) != 0:
                n = len(q)
                for p in range(n):
                    head = q.pop(0)
                    L = M[head]
                    for idx in range(len(L)):
                        if L[idx] == 1:
                            if idx not in visited:
                                visited.append(idx)
                                q.append(idx)
            res += 1            
        return res
                    